ping-payload-extensions/ping-custom-payload-extensions.md

```diff
@@ -0,0 +1,104 @@
+# Ping Custom Payload Extensions
+
+## Motivation
+
+Ping Payload Extensions. There is a minimal set of `standard extensions` which require a fork hard fork to change. This framework allows Portal clients to implement `non standard extensions` which don't require a hard fork to deploy to the network. A more flexible way to extend the Protocol without bloating the [Portal-Wire-Protocol](../portal-wire-protocol.md)
+
+# Type's
+
+There are 65536 unique type ids.
+
+Types 0-10_000 and 65436-65535 are reserved for for future upgrades.
+
+The rest are first come first serve, but they should still be defined in this repo to avoid overlaps.
+
+
+## Requirements
+
+All payloads used in the Ping `custom_payload` MUST follow the `Ping Custom Payload Extensions` format.
+
+## Custom Payload Extensions Format
+
+- **type**: numeric identifier which tells clients how the `payload` field should be decoded.
+- **payload**: the SSZ encoded extension payload
+
+
+```python
+CustomPayloadExtensionsFormat = Container(
+  type: u16,
+  payload: ByteList[max_length=1100]
+)
+```
+
+## Ping vs Pong
+The relationship between Ping and Pong message will be determined by the requirements of the type.
+
+Currently type 1,2,3 are symmetrical, having the same payload for both request and response.  This symmetry is not required.  Extensions may define separate payloads for Ping and Pong within the same type.
+
+
+### Error responses
+If the ping receiver can't handle the ping for any reason the pong should return an error payload
+
+Pong payload
+```python
+
+# Max ASCII hex encoded strings length
+MAX_ERROR_BYTE_LENGTH = 300
+
+error_payload = SSZ.serialize(Container(error_code: u16, message: ByteList[MAX_ERROR_BYTE_LENGTH]))
+
+ErrorPayload = Container(
+  type: 65535,
+  payload: error_payload
+)
+```
+
+### Error Code's
+
+#### 0: Extension not supported
+This code should be returned if the extension isn't supported. This error should only be returned if 
+- The extension isn't supported
+- The extension isn't a required extension for specified Portal Network.
+
+#### 1: Requested data not found
+This error code should be used when a client is unable to provide the necessary data for the response.
+
+#### 2: Failed to decode payload
+Wasn't able to decode the payload
+
+#### 3: System error
+A critical error happened and the ping can't be processed
+
+## Standard extensions
+
+Every Portal Network Client is required to support 3 extensions
+- [Type 0: Capabilities Payload](extensions/type-0.md): useful for finding what ping extensions a client supports
+- Sub-network standard extension. For history this is currently [Type 2: History Radius Payload](extensions/type-2.md). For State and Beacon the base extension is [Type 1: Basic Radius Payload](extensions/type-1.md). These are extensions what are crucial. This standard extensions are specified the respective sub-networks specification  
+- Type 3: [Get Client info Payload](extensions/type-3.md)
```

Comment:
I suggest we remove this from the standard set.  Clients *should* support it but I don't think it makes sense to make it be required.

Reply:
I think it is fine to make it so a valid option is an empty string for privacy inclined users, but I don't think we should remove it, as 99.9% of honest Portal nodes would support it. ` Client Info` is important for 
- getting better statistics of our network, 
  - useful for glados
  - tells us what platforms our popular and we should support better e.x. x86/arm/riscv5
- allows us to better track what client is creating bugs on the network. Are they running an old commit? Is it a CPU architecture specific issue? etc

If this isn't a default extension it would require 
- one ping to get `capabilities`
- one ping to get `client info`

If 99.9% of honest clients are going to support this call, I think it should be a standard extension, with an opt out to return an empty string by the 0.001% of users which have a privacy concern.

Reply:
I think I agree with this.  Will set the default path towards clients supporting this which I think is positive for network monitoring/health.

Reply:
If it is fine to just be an empty string, then it is a bit silly to send this around as another message req/rep.

Perhaps just add it as a field in the capabilities extension?

Reply:
I will update the PR to make it so client info is a part of the capabilities extension, this will limit the max amount of ping extensions a client can have enabled at a time to 450 or 400 instead of 500 to reserve byte space for the client enough, but 400 extensions should be more then enough so it isn't a deal breaker at the end of the day.

Reply:
To be honest, I don't think anybody would want to be writing Portal implementations if they have to deal with 400 (or even 40 for that matter) ping extensions :).